libs/common/src/change-stream/delete-many.ts

```diff
@@ -0,0 +1,72 @@
+import 'dotenv/config.js';
+import type { Filter, TransactionOptions, WithoutId } from 'mongodb';
+import { add } from 'date-fns';
+import type { AuditDetails, DeletionAuditLog, MongoDbCollectionName } from '../types';
+import { MongoDbClient } from '../mongo-db-client';
+import { generateAuditDatabaseRecordFromAuditDetails } from './generate-audit-database-record';
+import { changeStreamConfig } from './config';
+
+const { DELETION_AUDIT_LOGS_TTL_SECONDS } = changeStreamConfig;
+
+type DeleteManyParams = {
+  filter: Filter<any>;
+  collectionName: MongoDbCollectionName;
+  db: MongoDbClient;
+  auditDetails: AuditDetails;
+};
+
+const deleteManyWithAuditLogs = async ({ filter, collectionName, db, auditDetails }: DeleteManyParams) => {
+  const client = await db.getClient();
+  const session = client.startSession();
+
+  try {
+    const transactionOptions: TransactionOptions = {
+      readConcern: { level: 'snapshot' },
+      writeConcern: { w: 'majority' },
+    };
+    await session.withTransaction(async () => {
+      const collection = await db.getCollection(collectionName);
+      const documentsToDelete = await collection.find(filter, { projection: { _id: true }, session }).toArray();
+
+      if (documentsToDelete.length) {
+        const logsToInsert: WithoutId<DeletionAuditLog>[] = documentsToDelete.map(({ _id }) => ({
+          collectionName,
+          deletedDocumentId: _id,
+          auditRecord: generateAuditDatabaseRecordFromAuditDetails(auditDetails),
+          expireAt: add(new Date(), { seconds: DELETION_AUDIT_LOGS_TTL_SECONDS }),
+        }));
+
+        const deletionCollection = await db.getCollection('deletion-audit-logs');
+        const insertResult = await deletionCollection.insertMany(logsToInsert, { session });
+        if (!insertResult.acknowledged) {
+          throw new Error('Failed to create deletion audit logs');
+        }
+
+        const deleteResult = await collection.deleteMany({ $or: documentsToDelete.map(({ _id }) => ({ _id })) }, { session });
+        if (!(deleteResult.acknowledged && deleteResult.deletedCount === documentsToDelete.length)) {
+          throw new Error('Failed to delete documents');
+        }
+      }
+    }, transactionOptions);
+  } catch (error) {
+    console.error(`Failed to delete many, rolling back changes. Deleting from collection ${collectionName} with filter ${JSON.stringify(filter)}`);
+    throw error;
+  } finally {
+    await session.endSession();
+  }
+};
+
+export const deleteMany = async ({ filter, collectionName, db, auditDetails }: DeleteManyParams): Promise<{ acknowledged: boolean }> => {
+  if (process.env.CHANGE_STREAM_ENABLED === 'true') {
+    await deleteManyWithAuditLogs({
+      filter,
+      collectionName,
+      db,
+      auditDetails,
+    });
+
+    return { acknowledged: true };
```

Comment:
Do we not want to return the `deletedCount` from `deleteManyWithAuditLogs`, which we could then return here?

Reply:
I thought about this but it'd require a little bit of tweaking & decided to go for YAGNI. Happy to do so if you think there's serious benefit here

Reply:
Happy to leave it as I don't see any particular benefit.